Bluetooth: Mesh: Only settings load post mesh init

Andrewpini opened the pull request:
Prevent mesh stored settings from triggering unless bt_mesh_init has been called.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

omkar3141 reviewed (commented):
This is probably the simplest solution. Additionally, I think you need to also gate `bt_mesh_start()` call from the `mesh_commit()` using a similar condition. Because, if the mesh is not initialized, the setting load process should not start dynamic behaviors for beacons, hb, models, gatt, etc.

PavelVPV commented on subsys/bluetooth/mesh/main.c:
```diff
@@ -41,6 +41,8 @@
 #include "mesh.h"
 #include "gatt_cli.h"
 
+static bool is_initialized;
```
This could be in `bt_mesh_net.flags`: https://github.com/zephyrproject-rtos/zephyr/blob/53e1ea58e04c445f314c70ced1c8aa6ea0b53ff8/subsys/bluetooth/mesh/net.h#L169-L185

PavelVPV commented on subsys/bluetooth/mesh/heartbeat.c:
```diff
@@ -409,6 +409,10 @@ void bt_mesh_hb_resume(void)
 static int hb_pub_set(const char *name, size_t len_rd,
 		      settings_read_cb read_cb, void *cb_arg)
 {
+	if (!bt_mesh_is_initialized()) {
```
I think this could be hidden under `BT_MESH_SETTINGS_DEFINE`, e.g.:
```
#define BT_MESH_SETTINGS_DEFINE(_hname, _subtree, _set)			     \
	static int pre_ ## _set(const char *name, size_t len_rd, \
		      settings_read_cb read_cb, void *cb_arg) \
	{ \
		if (!atomic_test_bit(bt_mesh.flags, BT_MESH_INIT)) { \
			return 0; \
		} \
		return _set(name, len_rd, read_cb, cb_arg); \
	} \
	SETTINGS_STATIC_HANDLER_DEFINE(bt_mesh_##_hname, "bt/mesh/" _subtree, \
				       NULL, pre_##_set, NULL, NULL)
```

PavelVPV replied:
However, I would really like to be sure that there is no more neat way of solving this. Have you checked if there is another way to make settings not load subtree? Or maybe we could let settings subsystem recall mesh settings but once we get to `mesh_commit`, check that stack was initialized. Though in this case we have to be sure that nothing dangerous happens in `_set` handlers.

Laczen replied:
A more neat way to solve this might be to pass all settings for bt/bt_mesh trough the bt settings handler, the generic bt handler can then disable all bt_mesh settings handlers. Possibly it could even avoid all bt_mesh settings handlers to be registered.

PavelVPV replied:
@Laczen could you elaborate on your suggestion? Are talking about calling all mesh set handlers from bt settings set handler instead of registering them through `SETTIGNS_STATIC_HANDLER_DEFINE`?

PavelVPV replied:
I tried your suggestion. I think because I have to parse subtree inside mesh, it adds little bit more footprint in comparison with flag check in every set handler (tested on `mesh_shell` sample), but this probably can be improved:

Default:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      369292 B         1 MB     35.22%
             RAM:       50508 B       256 KB     19.27%
        IDT_LIST:          0 GB         2 KB      0.00%
```

With pre_##_set:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      369532 B         1 MB     35.24%
             RAM:       50508 B       256 KB     19.27%
        IDT_LIST:          0 GB         2 KB      0.00%
```

With own iterable section (https://github.com/PavelVPV/zephyr/commit/860214998860d839846221f659121499cfbdf24b):
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      369744 B         1 MB     35.26%
             RAM:       53644 B       256 KB     20.46%
        IDT_LIST:          0 GB         2 KB      0.00%
```

omkar3141 replied:
It will be so much better if the settings_load_subtree() API has another parameter that accepts array of subtree string pointers (acting as filters) that can be included/excluded from loading. Something like:
`int settings_load_subtree(const char *subtree, const char *subtree_filter_list, size_t filter_entries, enum filter_type_t filter_type);`
Where filter_type_t, can specify the inclusion or exclusion filter.

Laczen replied:
Cool, there seems to be quite some higher ram usage. I am sure this can be improved, but at least you have shown how it could be done without large code modification.

Laczen replied:
@omkar3141, that wouldn't be a solution as it is a call to `settings_load()` that is generating this problem. If all systems would limit themselves to a `settings_load_subtree()` of their own subtree the problem would also disappear.

omkar3141 replied:
One could remove the call to settings_load() from the application and instead directly use settings_load_subtree() as a top-level loading function. However, there can be application-specific concerns (such as this issue) that may require loading only parts of the tree, for example, when separate subsystems are present and only some of them needed to be active. Pushing these calls in the sub-system init sequence is also not an option because that makes the user application inflexible. @PavelVPV's proposal seems like a good compromise for now.

Laczen replied:
@PavelVPV, I think your `pre ## set` solution is the nicest. It could even be extended to taking a test condition as part of the macro (as a future settings improvement).

Maybe handling for a NULL _set routine should be added at the macro level.

Andrewpini commented:
From what I can see your suggestion works @PavelVPV. You think it is worth our time to do further exploration on how to do this more neat/efficient, or are we content with this solution?

PavelVPV commented:
Should we check `BT_MESH_INIT` flags in `bt_mesh_provision`? E.g. in case when `BT_SETTINGS` disabled and an application self-provisions the device?

alxelax commented:
> Should we check `BT_MESH_INIT` flags in `bt_mesh_provision`? E.g. in case when `BT_SETTINGS` disabled and an application self-provisions the device?

I guess it is not necessary since it is already checked implicitly here: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/settings.c#L64

If initialization happened then `BT_MESH_VALID` is set that is checked in the provisioning procedure.

PavelVPV commented:
> > Should we check `BT_MESH_INIT` flags in `bt_mesh_provision`? E.g. in case when `BT_SETTINGS` disabled and an application self-provisions the device?
> 
> I guess it is not necessary since it is already checked implicitly here: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/settings.c#L64
> 
> If initialization happened then `BT_MESH_VALID` is set that is checked in the provisioning procedure.

In case `BT_SETTINGS` is disabled, settings.c is not compiled.
See this: https://github.com/zephyrproject-rtos/zephyr/blob/0620cb1fe1807a4a0e1ea83b300138d27b7fc175/subsys/bluetooth/mesh/CMakeLists.txt#L27

alxelax commented:
@Andrewpini, seems these changes broke oob provisioning bsim tests.

Andrewpini committed:
Bluetooth: Mesh: Only settings load post mesh init

Prevent mesh stored settings from triggering unless bt_mesh_init
has been called.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

alxelax commented on subsys/bluetooth/mesh/settings.c:
```diff
@@ -63,6 +63,10 @@ int bt_mesh_settings_set(settings_read_cb read_cb, void *cb_arg,
 
 static int mesh_commit(void)
 {
+	if (!atomic_test_bit(bt_mesh.flags, BT_MESH_INIT)) {
```
Isn't `mesh_commit` just one more `pre_set` function that already has this condition?

Andrewpini replied:
It is defined a little bit differently than the other ones. Not sure if it can be handled in a simmilar manner or if it needs to stay as it is.

alxelax replied:
I see it uses another macro. Probably it is better to use the same as the others but it is out of scope of this PR.

PavelVPV replied:
> Probably it is better to use the same as the others but it is out of scope of this PR.

This is intentional. E.g `BT_MESH_SETTINGS_DEFINE` doesn't have `commit` handler on purpose.
